File: 22FALL/282.expression-add-operators.py
#
# @lc app=leetcode id=282 lang=python3
#
# [282] Expression Add Operators
#

# @lc code=start
class Solution:
    def addOperators(self, num: str, target: int) -> List[str]:
        self.ans = []
        self.res = 0
        self.target = target
        self.num = num
        self.n = len(num)
        self.backtrack([], 0, 0, 0)
        return self.ans

    def backtrack(self, expr, i, cur_val, cur_mul):
        if i == self.n:
            if cur_val == self.target:
                self.ans.append(''.join(expr))
            return
        sign_index = len(expr)
        if i > 0:
            expr.append('')
        val = 0
        for j in range(i, self.n):
            if (j>i and self.num[i]=='0'):
                break
            val = 10 * val + int(self.num[j])
            expr.append(self.num[j])
            if i==0:
                self.backtrack(expr, j+1, val, val)
            else:
                expr[sign_index] = '+'
                self.backtrack(expr, j+1, cur_val+val, val) 
                expr[sign_index] = '-'
                self.backtrack(expr, j+1, cur_val-val, -val) 
                expr[sign_index] = '*'
                self.backtrack(expr, j+1, cur_val-cur_mul + val*cur_mul, cur_mul*val)
        del expr[sign_index:]
# @lc code=end

File: Java/leetcode/editor/en/282.expression-add-operators.java
import java.util.ArrayList;
import java.util.List;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    List<String> ans;
    int n;
    int target;
    String num;

    public void backTrack(StringBuffer expr, int i, long res, long mul){
        if (i==n){
            if(res==target){
                ans.add(expr.toString());
            }
            return;
        }
        int signIndex = expr.length();
        if(i>0){
            expr.append(0);
        }
        long val = 0;
        for (int j=i; j<n && (j==i || num.charAt(i)!='0'); ++j) {
            val = val * 10 + num.charAt(j)-'0';
            expr.append(num.charAt(j));
            if(i==0){
                backTrack(expr, j+1, val, val);
            } else {
                expr.setCharAt(signIndex, '+');
                backTrack(expr,j+1,res+val, val);
                expr.setCharAt(signIndex, '-');
                backTrack(expr, j+1, res-val, -val);
                expr.setCharAt(signIndex, '*');
                backTrack(expr, j+1, res-mul + mul*val, mul*val);
            }
        }
        expr.setLength(signIndex);
    }

    public List<String> addOperators(String num, int target) {
        this.n = num.length();
        this.num = num;
        this.target = target;
        this.ans = new ArrayList<String>();
        StringBuffer expr = new StringBuffer();
        backTrack(expr, 0, 0, 0);
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
